plasmapy/atomic/particle_input.py

```diff
@@ -0,0 +1,241 @@
+import functools
+import inspect
+
+from .particle_class import Particle
+from ..utils import (AtomicError,
+                     InvalidParticleError,
+                     InvalidElementError,
+                     InvalidIonError,
+                     InvalidIsotopeError,
+                     ChargeError)
+
+from typing import Callable, Union, Any, Set, List, Tuple
+
+# TODO: change any keyword to any_of and allow it to be a list
+
+
+def particle_input(wrapped_function: Callable = None,
+                   must_be: Union[str, Set, List, Tuple] = set(),
+                   any_of: Union[str, Set, List, Tuple] = set(),
+                   exclude: Union[str, Set, List, Tuple] = set(),
+                   none_shall_pass: bool = False,
+                   **kwargs) -> Any:
+    r"""A decorator to take arguments and keywords related to a particle
+    and pass through the Particle class to the callable instead.
+
+    Parameters
+    ----------
+
+    wrapped_function : callable
+        The function to be decorated.
+
+    must_be : str, set, list, or tuple; optional
+        A list of categories
+
+    any_of : str, set, list, or tuple; optional
+
+    exclude : str, set, list, or tuple; optional
+
+    none_shall_pass : bool
+        If set to True, then the decorated argument is allowed to be set
+        to None without raising an exception.  In such cases, this
+        decorator will pass through None to the decorated function.
+
+    Notes
+    -----
+
+    This version of particle_input only works with functions, but should
+    be extended to work with classes and methods.
+
+    """
+
+    def _particle_errmsg(argname: str, argval: str, Z: int = None,
```

Comment:
Do these helper functions need to be defined inside the decorator? I feel like it would be neater if they were outside.

Reply:
I had been debating this too, so I'll define these outside.